### PR TITLE
feat(Shiroko): add Presence

### PR DESCRIPTION
**Pull request by kaisei-kto**
## Description 

Add Presence for [Shiroko](https://shiroko.co/), which is a service for streaming anime.

## Acknowledgements
- [x] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [x] I linted the code by running `yarn format`
- [x] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

image
image
image
image

</details>

### CODE REVIEW

File: package-lock.json
Comment (darkvillager2) [inline package-lock.json:1]: Remove changes to this.

File: websites/S/Shiroko/metadata.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+		"id": "84719369069805568",
+		"name": "laziestboy"
+	},
+	"service": "Shiroko",
+	"description": {
+		"en": "A website to watch anime online for free with no ads."
+	},
+	"url": "shiroko.co",
+	"version": "1.0.0",
+	"logo": "https://i.postimg.cc/1zN0SNrb/shiroko-vector-1.png",
```
Comment (darkvillager2) [inline websites/S/Shiroko/metadata.json:14]: Logo must be 512x512

File: websites/S/Shiroko/presence.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	startTimestamp = Math.floor(Date.now() / 1e3),
+	lastHref: string,
+	lastUpdate = 0;
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			buttons: [{ label: "Visit Shiroko", url: "https://shiroko.co" }],
+			startTimestamp,
+		},
+		video = document.body.querySelectorAll("video")[0];
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:19]:
```suggestion
		video = document.querySelector("video");
```

File: websites/S/Shiroko/presence.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	playback = false,
+	startTimestamp = Math.floor(Date.now() / 1e3),
+	lastHref: string,
+	lastUpdate = 0;
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			buttons: [{ label: "Visit Shiroko", url: "https://shiroko.co" }],
+			startTimestamp,
+		},
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:18]:
```suggestion
		},
	{ pathname, href } = document.location;
```

File: websites/S/Shiroko/presence.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		presenceData.smallImageKey = playback ? Assets.Play : Assets.Pause;
+		presenceData.smallImageText = playback ? "Playing" : "Paused";
+	}
+
+	if (video && lastPlaybackState === playback) return;
+	lastPlaybackState = playback;
+
+	presenceData.endTimestamp = null;
+
+	const { pathname, href } = location;
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:37]:
```suggestion
```

File: websites/S/Shiroko/presence.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+	if (pathname.startsWith("/en/schedule")) {
+		presenceData.details = "Viewing Schedule...";
+		presenceData.buttons.unshift({
+			label: "View Schedule",
+			url: `${DOMAIN}/en/schedule`,
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:43]:
```suggestion
			url: href
```

File: websites/S/Shiroko/presence.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+		presenceData.details = "Searching Manga...";
+	if (pathname.includes("/search/anime"))
+		presenceData.details = "Searching Anime...";
+	if (href.includes("/search/anime?season="))
+		presenceData.details = "Viewing Season...";
+	if (
+		pathname.startsWith("/en/anime/") &&
+		!pathname.startsWith("/en/anime/watch")
+	) {
+		presenceData.details = `${getTitle()}`;
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:56]:
```suggestion
		presenceData.details = getTitle()
```

File: websites/S/Shiroko/utils.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+export function getEpisode(): number {
+	let episode = 1;
+	const query = document.querySelector(
+		"#primary > div.flex.flex-col.px-3.lg\\:px-0.mt-2 > div.flex.flex-col.gap-2.font-karla.pt-2.pb-3.w-full > div.flex.justify-between.items-center > div:nth-child(1) > div"
+	);
+
+	if (query) episode = Number(query.children[0].innerHTML.split(" ").pop());
+
+	return episode;
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:39]:
```suggestion
	const query = [...document.querySelectorAll("h3")]
   .filter(a => a.textContent.includes("Episode"))[0]?.textContent?.replace("Episode ", "")
	if (query) episode = query

	return episode;
```

File: websites/S/Shiroko/utils.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+			["poster anime", "Poster Anime", "Anime Cover"].includes(
+				String(img.attributes.getNamedItem("alt")?.value)
+			)
+		)
+			return String(img.src);
+	}
+}
+
+export function getTitle(): string | undefined {
+	const nt = document.querySelectorAll(".title-nt")[0],
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:13]:
```suggestion
	const nt = document.querySelector(".title-nt"),
```

File: websites/S/Shiroko/utils.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		ro = document.querySelectorAll(".title-rm")[0],
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:14]:
```suggestion
		ro = document.querySelector(".title-rm"),
```

File: websites/S/Shiroko/utils.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+		en = document.querySelectorAll(".title-en")[0];
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:15]:
```suggestion
		en = document.querySelector(".title-en");
```

File: websites/S/Shiroko/utils.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+	if (en) return en.innerHTML;
+	if (ro) return ro.innerHTML;
+	if (nt) return nt.innerHTML;
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:19]:
```suggestion
	if (en) return en.textContent;
	if (ro) return ro.textContent;
	if (nt) return nt.textContent;
```

File: websites/S/Shiroko/utils.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+	for (const div of document.querySelectorAll("div")) {
+		if (
+			div.className === "flex flex-col gap-1 text-center md:text-start w-full"
+		) {
+			for (const element of div.children)
+				if (element.tagName.toLowerCase() === "h1") return element.innerHTML;
+		}
+	}
+}
```
Comment (darkvillager2) [inline websites/S/Shiroko/utils.ts:29]: See filter query requested earlier, maybe use that here.

File: websites/S/Shiroko/presence.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
+import { getEpisode, getId, getImage, getTitle } from "./utils";
+
+const presence = new Presence({
+		clientId: "1286500193881161804",
+	}),
+	DOMAIN = "https://shiroko.co";
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:6]: Not needed anymore?

File: websites/S/Shiroko/presence.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+		presenceData.details = `${getTitle()}`;
+		presenceData.state = !isNaN(episode)
+			? `Episode ${episode}`
+			: "Unable to retrieve episode";
+		presenceData.type = ActivityType.Watching;
+
+		presenceData.buttons.unshift({
+			label: "Watch Anime",
+			url: `${DOMAIN}/en/anime/watch?id=${getId()}&n=${episode}`,
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:85]:
```suggestion
			url: href
```

File: websites/S/Shiroko/presence.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+	if (pathname.startsWith("/en/profile/")) {
+		const split = pathname.split("/").slice(1);
+		if (split.length === 3) {
+			const username = split.pop();
+
+			presenceData.details = `Viewing ${username}'s Profile`;
+			presenceData.buttons.unshift({
+				label: "View Profile",
+				url: `${DOMAIN}/en/profile/${username}`,
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:97]:
```suggestion
				url: href
```

**Review by darkvillager2 — CHANGES_REQUESTED** [review]
Hii, thanks for your contribution.

I have some changes i'd like to see. 

If you have any questions regarding my requested changes feel free to dm me on discord (@Dark_Ville) or ask in this PR.

File: websites/S/Shiroko/presence.ts [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+	) {
+		presenceData.details = `${getTitle()}`;
+		presenceData.state = "Viewing Anime...";
+
+		presenceData.buttons.unshift({
+			label: "View Anime",
+			url: `${DOMAIN}/en/anime/${getId()}`,
+		});
+	} else if (pathname.startsWith("/en/manga/")) {
+		presenceData.details = `${getTitle()}`;
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:64]:
```suggestion
		presenceData.details = getTitle()
```

File: websites/S/Shiroko/presence.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+		presenceData.state = "Viewing Manga...";
+
+		presenceData.buttons.unshift({
+			label: "View Manga",
+			url: `${DOMAIN}/en/manga/${getId()}`,
```
Comment (darkvillager2) [inline websites/S/Shiroko/presence.ts:69]:
```suggestion
			url: refh
```

File: websites/S/Shiroko/metadata.json [hunk 18]
```diff
@@ -0,0 +1,22 @@
+{
+	"$schema": "https://schemas.premid.app/metadata/1.11",
+	"apiVersion": 1,
+	"author": {
+		"id": "84719369069805568",
+		"name": "laziestboy"
+	},
+	"service": "Shiroko",
+	"description": {
+		"en": "A website to watch anime online for free with no ads."
```
Comment (darkvillager2) [inline websites/S/Shiroko/metadata.json:10]:
```suggestion
		"en": "Discover your new favorite anime or manga title! Shiroko offers a vast library of high-quality content, accessible on multiple devices and without any interruptions. Start using Shiroko today!"
```

**Review by darkvillager2 — COMMENTED** [review]
other than this, LGTM

**Review by Bas950 — APPROVED** [review]

File: websites/S/Shiroko/presence.ts [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+
+let lastPlaybackState = false,
+	playback = false,
+	startTimestamp = Math.floor(Date.now() / 1e3),
+	lastHref: string,
+	lastUpdate = 0;
+
+presence.on("UpdateData", async () => {
+	const presenceData: PresenceData = {
+			buttons: [{ label: "Visit Shiroko", url: "https://shiroko.co" }],
```
Comment (theusaf) [inline websites/S/Shiroko/presence.ts:15]: Buttons which direct to the main page are [prohibited](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md#presencets).
```suggestion
```

File: websites/S/Shiroko/presence.ts [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+			startTimestamp = Math.floor(Date.now() / 1000);
+
+		presenceData.smallImageKey = playback ? Assets.Play : Assets.Pause;
+		presenceData.smallImageText = playback ? "Playing" : "Paused";
+	}
+
+	if (video && lastPlaybackState === playback) return;
+	lastPlaybackState = playback;
+
+	presenceData.endTimestamp = null;
```
Comment (theusaf) [inline websites/S/Shiroko/presence.ts:35]:
```suggestion
	delete presenceData.endTimestamp;
```

File: websites/S/Shiroko/presence.ts [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+	if (pathname.startsWith("/en/schedule")) {
+		presenceData.details = "Viewing Schedule...";
+		presenceData.buttons.unshift({
+			label: "View Schedule",
+			url: href,
+		});
```
Comment (theusaf) [inline websites/S/Shiroko/presence.ts:42]: After the previous requested changes, these will have to be adjusted

File: websites/S/Shiroko/presence.ts [hunk 22]
```diff
@@ -0,0 +1,107 @@
...
+			presenceData.details = `Viewing ${username}'s Profile`;
+			presenceData.buttons.unshift({
+				label: "View Profile",
+				url: href,
+			});
+		}
+	}
+
+	const now = Date.now();
+	if (lastHref !== href || (now - lastUpdate) / 1e3 >= 1) {
```
Comment (theusaf) [inline websites/S/Shiroko/presence.ts:99]:
```suggestion
	if (lastHref !== href || (now - lastUpdate) / 1000 >= 1) {
```

File: websites/S/Shiroko/utils.ts [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+		.find(a => a.textContent.includes("Episode"))
+		?.textContent?.replace("Episode ", "");
+
+	if (query) episode = Number(query);
+
+	return episode;
+}
+
+export function getId(): number {
+	for (const [key, value] of location.search
```
Comment (theusaf) [inline websites/S/Shiroko/utils.ts:36]:
```suggestion
	for (const [key, value] of document.location.search
```

File: websites/S/Shiroko/utils.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+		.substr(1)
+		.split("&")
+		.map(s => s.split("=")))
+		if (key === "id") return Number(value);
+
+	for (const path of location.pathname.split("/"))
```
Comment (theusaf) [inline websites/S/Shiroko/utils.ts:42]:
```suggestion
	for (const path of documenet.location.pathname.split("/"))
```

**Review by theusaf — CHANGES_REQUESTED** [review]

File: websites/S/Shiroko/utils.ts [hunk 25]
```diff
@@ -33,12 +33,14 @@ export function getEpisode(): number {
 }
 
 export function getId(): number {
-	for (const [key, value] of location.search
+	const { search, pathname } = document.location;
```
[the remark this answers is not in the document]
Comment (theusaf) [reply websites/S/Shiroko/utils.ts:36]: That's fine

**Review by theusaf — APPROVED** [review]

**Pull request merged**